paho/client.go

```diff
@@ -127,154 +138,167 @@ func NewClient(conf ClientConfig) *Client {
 // returned. Otherwise the failure Connack (if there is one) is returned
 // along with an error indicating the reason for the failure to connect.
 func (c *Client) Connect(ctx context.Context, cp *Connect) (*Connack, error) {
-	cleanup := func() {
-		select {
-		case <-c.stop:
-			//already shutting down, do nothing
-		default:
-			close(c.stop)
-		}
-		c.Conn.Close()
-	}
 	if c.Conn == nil {
 		return nil, fmt.Errorf("client connection is nil")
 	}
+	c.connectOnce.Do(func() {
```

Comment:
Can you please give me a bit of background about this change? Is calling `Connect` multiple times needed thus the `connectOnce`?

Reply:
It is vice versa: Connect() must be called, and only one call is expected. So this `sync.Once` protects the unexported fields which are populated during connection handshake. All other methods expect the Connect() was called before.

Reply:
Do we really need to protect it then? Can't we just call `Connect` once? How are handling reconnections btw, is this related?

Reply:
Yes, if you will not protect them, then the second call would change the fields, which are accessed everywhere without mutex. That is, it would be a race.

Reconnections are handled by our wrapper.

Reply:
Ouch. Alright then, I don't see a reason for not doing this now especially since we have our own wrapper around it. Hopefully we'll have the time to address this properly in the future :+1: